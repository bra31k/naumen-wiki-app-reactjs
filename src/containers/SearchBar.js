import React, { Component } from 'react';
import { connect } from "react-redux"
import PropTypes from "prop-types"

import { getArticles } from '../actions/articlesActions'



export class SearchBar extends Component {

    constructor(props) {

        super(props)
        this.state = {
            inputValue: "",
            searchValue: "",
            sort: "relevance",
            sroffset: 0,
            suggestion: [],
            shouldSuggestion: false,
            theme: 'light',
        }

        this.handleChangeSelect = this.handleChangeSelect.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
        this.suggestionClick = this.suggestionClick.bind(this)
        this.toggleTheme = this.toggleTheme.bind(this)
        this.handleChangeInput = this.handleChangeInput.bind(this)
    }

    componentDidMount() {
        if (localStorage.getItem('naumen-wiki-app')) {
            let suggestion = JSON.parse(localStorage.getItem('naumen-wiki-app'))
            suggestion.forEach(sug => {
                this.setState(prevState => ({
                    suggestion: [...prevState.suggestion, sug]
                }))
            })
        }
        window.addEventListener('scroll', this.onScroll, false);
    }

    componentDidUpdate(prevProps, prevState) {
        if ((prevState.searchValue !== this.state.searchValue
            || prevState.sort !== this.state.sort
            || prevState.sroffset !== this.state.sroffset )
            && this.state.searchValue.length > 0)
        {
            this.props.getArticles(this.state.searchValue, this.state.sort, this.state.sroffset)
            localStorage.setItem('naumen-wiki-app', JSON.stringify(this.state.suggestion))
        }
    }

    componentWillUnmount() {
        window.removeEventListener('scroll', this.onScroll, false);
    }

    suggestionClick = (value) => {
        this.setState({
            inputValue: value,
            searchValue: value,
            sroffset: 0,
        })
    }

    handleChangeSelect = (event) => {
        this.setState({
            sort: event.target.value,
            sroffset: 0,
        })
    }

    handleSubmit = (event) => {
        event.preventDefault()
        const inputValue = this.state.inputValue
        this.setState({
            searchValue: inputValue,
            sroffset: 0,
        })
        if (!this.state.suggestion.includes(inputValue) && inputValue.length>0)
        {
            if (this.state.suggestion.length >= 3) {
                this.setState({
                    suggestion: this.state.suggestion.splice(1),
                })
            }
            this.setState(prevState => ({
                suggestion: [...prevState.suggestion, inputValue],
            }))
        }

    }

    handleChangeInput = (event) => {
        this.setState({
            inputValue: event.target.value
        });
    }

    toggleTheme = () => {
        const theme = this.state.theme === 'dark' ? 'light' : 'dark';
        this.setState({
            theme: theme
        });
        document.documentElement.setAttribute("data-theme", theme);
    }

    onScroll = () => {
        const scrollHeight = Math.max(
            document.body.scrollHeight, document.documentElement.scrollHeight,
            document.body.offsetHeight, document.documentElement.offsetHeight,
            document.body.clientHeight, document.documentElement.clientHeight
        );
        if ((document.documentElement.clientHeight + window.scrollY) >= (scrollHeight)) {
            this.setState({
                sroffset: this.state.sroffset + 10,
            })
        }
    }


    render() {
        return (
            <div>
                <form onSubmit={(event) => this.handleSubmit(event)}>
                    <input
                        type="text"
                        id="search"
                        placeholder="Искать здесь..."
                        autoComplete="off"
                        value={this.state.inputValue}
                        onChange={this.handleChangeInput}
                    />
                        {this.state.inputValue.length > 0
                            ? <ul>{this.state.suggestion.map(sug => (
                            <li key={this.state.suggestion.indexOf(sug)}
                                onClick={() => this.suggestionClick(sug)}>
                                {sug}
                            </li>
                            ))}</ul>
                            : null}
                        <button type="submit" className="button"/>
                </form>
                <select value={this.state.sort} onChange={(event) => this.handleChangeSelect(event)}>
                    <option value="relevance">Relevance</option>
                    <option value="just_match">Just match</option>
                    <option value="none">None</option>
                    <option value="incoming_links_asc">Incoming links asc</option>
                    <option value="incoming_links_desc">Incoming links esc</option>
                    <option value="last_edit_asc">Last edit asc</option>
                    <option value="last_edit_desc">Last edit desc</option>
                    <option value="create_timestamp_asc">Create timestamp asc</option>
                    <option value="create_timestamp_desc">Create timestamp desc</option>
                </select>
                <button id='button-change-theme' className="button-change-theme" onClick={this.toggleTheme}>Change theme</button>
            </div>
        );
    }
}

SearchBar.propTypes = {
    getArticles: PropTypes.func,
};

const mapDispatchToProps = dispatch => {
    return {
        getArticles: (inputValue, sort, sroffset) => dispatch(getArticles(inputValue, sort, sroffset)),
    }
}

export default connect(null, mapDispatchToProps)(SearchBar)