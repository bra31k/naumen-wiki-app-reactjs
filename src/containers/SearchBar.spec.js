import React from 'react'
import { SearchBar } from './SearchBar'
import { shallow } from 'enzyme'

describe('SearchBar', () => {

    const getArticlesMock = jest.fn()

    const props = {
        getArticles: getArticlesMock
    }

    describe('SearchBar initial', () => {
        const searchBar = shallow(<SearchBar {...props} />)

        it('SearchBar properly', () => {
            expect(searchBar).toMatchSnapshot()
        })
    })
})