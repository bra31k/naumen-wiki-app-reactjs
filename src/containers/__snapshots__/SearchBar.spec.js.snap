// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar SearchBar initial SearchBar properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBar
    getArticles={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
        >
          <input
            autoComplete="off"
            name="searchWiki"
            onChange={[Function]}
            placeholder="Искать здесь..."
            type="text"
            value=""
          />
          <button
            className="button"
            type="submit"
          />
        </form>,
        <select
          onChange={[Function]}
          value="relevance"
        >
          <option
            value="relevance"
          >
            Relevance
          </option>
          <option
            value="just_match"
          >
            Just match
          </option>
          <option
            value="none"
          >
            None
          </option>
          <option
            value="incoming_links_asc"
          >
            Incoming links asc
          </option>
          <option
            value="incoming_links_desc"
          >
            Incoming links esc
          </option>
          <option
            value="last_edit_asc"
          >
            Last edit asc
          </option>
          <option
            value="last_edit_desc"
          >
            Last edit desc
          </option>
          <option
            value="create_timestamp_asc"
          >
            Create timestamp asc
          </option>
          <option
            value="create_timestamp_desc"
          >
            Create timestamp desc
          </option>
        </select>,
        <button
          className="button-change-theme"
          onClick={[Function]}
        >
          Change theme
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              autoComplete="off"
              name="searchWiki"
              onChange={[Function]}
              placeholder="Искать здесь..."
              type="text"
              value=""
            />,
            null,
            <button
              className="button"
              type="submit"
            />,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "autoComplete": "off",
              "name": "searchWiki",
              "onChange": [Function],
              "placeholder": "Искать здесь...",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "button",
              "type": "submit",
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value="relevance"
            >
              Relevance
            </option>,
            <option
              value="just_match"
            >
              Just match
            </option>,
            <option
              value="none"
            >
              None
            </option>,
            <option
              value="incoming_links_asc"
            >
              Incoming links asc
            </option>,
            <option
              value="incoming_links_desc"
            >
              Incoming links esc
            </option>,
            <option
              value="last_edit_asc"
            >
              Last edit asc
            </option>,
            <option
              value="last_edit_desc"
            >
              Last edit desc
            </option>,
            <option
              value="create_timestamp_asc"
            >
              Create timestamp asc
            </option>,
            <option
              value="create_timestamp_desc"
            >
              Create timestamp desc
            </option>,
          ],
          "onChange": [Function],
          "value": "relevance",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Relevance",
              "value": "relevance",
            },
            "ref": null,
            "rendered": "Relevance",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Just match",
              "value": "just_match",
            },
            "ref": null,
            "rendered": "Just match",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "None",
              "value": "none",
            },
            "ref": null,
            "rendered": "None",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Incoming links asc",
              "value": "incoming_links_asc",
            },
            "ref": null,
            "rendered": "Incoming links asc",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Incoming links esc",
              "value": "incoming_links_desc",
            },
            "ref": null,
            "rendered": "Incoming links esc",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Last edit asc",
              "value": "last_edit_asc",
            },
            "ref": null,
            "rendered": "Last edit asc",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Last edit desc",
              "value": "last_edit_desc",
            },
            "ref": null,
            "rendered": "Last edit desc",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Create timestamp asc",
              "value": "create_timestamp_asc",
            },
            "ref": null,
            "rendered": "Create timestamp asc",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Create timestamp desc",
              "value": "create_timestamp_desc",
            },
            "ref": null,
            "rendered": "Create timestamp desc",
            "type": "option",
          },
        ],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Change theme",
          "className": "button-change-theme",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Change theme",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
          >
            <input
              autoComplete="off"
              name="searchWiki"
              onChange={[Function]}
              placeholder="Искать здесь..."
              type="text"
              value=""
            />
            <button
              className="button"
              type="submit"
            />
          </form>,
          <select
            onChange={[Function]}
            value="relevance"
          >
            <option
              value="relevance"
            >
              Relevance
            </option>
            <option
              value="just_match"
            >
              Just match
            </option>
            <option
              value="none"
            >
              None
            </option>
            <option
              value="incoming_links_asc"
            >
              Incoming links asc
            </option>
            <option
              value="incoming_links_desc"
            >
              Incoming links esc
            </option>
            <option
              value="last_edit_asc"
            >
              Last edit asc
            </option>
            <option
              value="last_edit_desc"
            >
              Last edit desc
            </option>
            <option
              value="create_timestamp_asc"
            >
              Create timestamp asc
            </option>
            <option
              value="create_timestamp_desc"
            >
              Create timestamp desc
            </option>
          </select>,
          <button
            className="button-change-theme"
            onClick={[Function]}
          >
            Change theme
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                autoComplete="off"
                name="searchWiki"
                onChange={[Function]}
                placeholder="Искать здесь..."
                type="text"
                value=""
              />,
              null,
              <button
                className="button"
                type="submit"
              />,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "autoComplete": "off",
                "name": "searchWiki",
                "onChange": [Function],
                "placeholder": "Искать здесь...",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            null,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "button",
                "type": "submit",
              },
              "ref": null,
              "rendered": null,
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                value="relevance"
              >
                Relevance
              </option>,
              <option
                value="just_match"
              >
                Just match
              </option>,
              <option
                value="none"
              >
                None
              </option>,
              <option
                value="incoming_links_asc"
              >
                Incoming links asc
              </option>,
              <option
                value="incoming_links_desc"
              >
                Incoming links esc
              </option>,
              <option
                value="last_edit_asc"
              >
                Last edit asc
              </option>,
              <option
                value="last_edit_desc"
              >
                Last edit desc
              </option>,
              <option
                value="create_timestamp_asc"
              >
                Create timestamp asc
              </option>,
              <option
                value="create_timestamp_desc"
              >
                Create timestamp desc
              </option>,
            ],
            "onChange": [Function],
            "value": "relevance",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Relevance",
                "value": "relevance",
              },
              "ref": null,
              "rendered": "Relevance",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Just match",
                "value": "just_match",
              },
              "ref": null,
              "rendered": "Just match",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "None",
                "value": "none",
              },
              "ref": null,
              "rendered": "None",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Incoming links asc",
                "value": "incoming_links_asc",
              },
              "ref": null,
              "rendered": "Incoming links asc",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Incoming links esc",
                "value": "incoming_links_desc",
              },
              "ref": null,
              "rendered": "Incoming links esc",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Last edit asc",
                "value": "last_edit_asc",
              },
              "ref": null,
              "rendered": "Last edit asc",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Last edit desc",
                "value": "last_edit_desc",
              },
              "ref": null,
              "rendered": "Last edit desc",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Create timestamp asc",
                "value": "create_timestamp_asc",
              },
              "ref": null,
              "rendered": "Create timestamp asc",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Create timestamp desc",
                "value": "create_timestamp_desc",
              },
              "ref": null,
              "rendered": "Create timestamp desc",
              "type": "option",
            },
          ],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Change theme",
            "className": "button-change-theme",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Change theme",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
